fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/AvailableFields.java

```diff
@@ -0,0 +1,211 @@
+                keyFields.addAll(groupingKeyExpression.getGroupingSubKey().normalizeKeyForPositions());
+            }
+        }
+
+        // Like FDBRecordStoreBase.indexEntryKey(), but with key expressions instead of actual values.
+        final List<KeyExpression> primaryKeys = commonPrimaryKey == null
+                                                ? Collections.emptyList()
+                                                : commonPrimaryKey.normalizeKeyForPositions();
+        index.trimPrimaryKey(primaryKeys);
+        keyFields.addAll(primaryKeys);
+
+        Map<KeyExpression, FieldData> fields = new HashMap<>();
+        final IndexKeyValueToPartialRecord.Builder builder = IndexKeyValueToPartialRecord.newBuilder(recordType);
+        for (int i = 0; i < keyFields.size(); i++) {
+            KeyExpression keyField = keyFields.get(i);
+            FieldData fieldData = FieldData.of(IndexKeyValueToPartialRecord.TupleSource.KEY, i);
+            if (!keyField.createsDuplicates() && addCoveringField(keyField, fieldData, builder)) {
+                fields.put(keyField, fieldData);
+            }
+        }
+        for (int i = 0; i < valueFields.size(); i++) {
+            KeyExpression valueField = valueFields.get(i);
+            FieldData fieldData = FieldData.of(IndexKeyValueToPartialRecord.TupleSource.VALUE, i);
+            if (!valueField.createsDuplicates() && addCoveringField(valueField, fieldData, builder)) {
+                fields.put(valueField, fieldData);
+            }
+        }
+
+        if (!builder.isValid()) {
+            return NO_FIELDS;
+        }
+        return new AvailableFields(fields);
+    }
+
+    public static boolean addCoveringField(@Nonnull KeyExpression requiredExpr,
+                                           @Nonnull FieldData fieldData,
+                                           @Nonnull IndexKeyValueToPartialRecord.Builder builder) {
+        while (requiredExpr instanceof NestingKeyExpression) {
+            NestingKeyExpression nesting = (NestingKeyExpression)requiredExpr;
+            String fieldName = nesting.getParent().getFieldName();
+            requiredExpr = nesting.getChild();
+            builder = builder.getFieldBuilder(fieldName);
+        }
+        if (requiredExpr instanceof FieldKeyExpression) {
+            FieldKeyExpression fieldKeyExpression = (FieldKeyExpression)requiredExpr;
+            if (!fieldKeyExpression.getNullStandin().equals(Key.Evaluated.NullStandin.NULL) ||
+                    fieldKeyExpression.getFanType().equals(KeyExpression.FanType.FanOut)) {
+                return false;
+            }
+            builder.addField(fieldKeyExpression.getFieldName(), fieldData.source, fieldData.index);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    @Nonnull
+    public static AvailableFields intersection(@Nonnull List<AvailableFields> toIntersect) {
```

Comment:
With regard to the first thought, I agree in principle but think this should be handled in a more systematic way. Right now, it's remarkably difficult to say exactly what fields are present, because the way that "present" is defined varies by context. For example, is a repeated field "present"? For the purposes of covering optimizations, it is not, but in general it probably is. I believe that the logic we have here is circumstantially correct and also fairly simple. The join-based framework you're working on will clarify a lot of these issues.

Regarding the second point, that's an oversight on my part. This is only ever called by union and intersection plans, w which are required to have a least two children. I've added a check that throws an exception if this gets called on an empty list.